src/core/include/units/magnitude.h

```diff
@@ -117,6 +118,98 @@ static constexpr bool is_base_power<base_power<T>> = true;
+  std::is_floating_point_v<T>,
+  long double,
+  std::conditional_t<std::is_signed_v<T>, std::intmax_t, std::uintmax_t>>;
+
+// Raise an arbitrary arithmetic type to a positive integer power at compile time.
+template<typename T> requires std::is_arithmetic_v<T>
+consteval T int_power(T base, std::integral auto exp){
+  if (exp < 0) { throw std::invalid_argument{"int_power only supports positive integer powers"}; }
+
+  // TODO(chogg): Unify this implementation with the one in pow.h.  That one takes its exponent as a
+  // template parameter, rather than a function parameter.
+
+  if (exp == 0) {
+    return 1;
+  }
+
+  if (exp % 2 == 1) {
+    return base * int_power(base, exp - 1);
+  }
+
+  const auto square_root = int_power(base, exp / 2);
+  const auto result = square_root * square_root;
+
+  if constexpr(std::is_unsigned_v<T>) {
+    // As this function can only be called at compile time, the exception functions as a
+    // "parameter-compatible static_assert", and does not result in exceptions at runtime.
+    if (result / square_root != square_root) { throw std::overflow_error{"Unsigned wraparound"}; }
+  }
+
+  return result;
+}
+
+
+template<typename T> requires std::is_arithmetic_v<T>
+consteval widen_t<T> compute_base_power(BasePower auto bp)
+{
+  // This utility can only handle integer powers.  To compute rational powers at compile time, we'll
+  // need to write a custom function.
+  //
+  // Note that since this function can only be called at compile time, the point of these exceptions
+  // is to act as "static_assert substitutes", not to throw actual exceptions at runtime.
+  if (bp.power.den != 1) { throw std::invalid_argument{"Rational powers not yet supported"}; }
+  if (bp.power.exp < 0) { throw std::invalid_argument{"Unsupported exp value"}; }
+
+  if (bp.power.num < 0) {
+    if constexpr (std::is_integral_v<T>) {
+      throw std::invalid_argument{"Cannot represent reciprocal as integer"};
+    } else {
+      return 1 / compute_base_power<T>(inverse(bp));
+    }
+  }
+
+  auto power = bp.power.num * int_power(10, bp.power.exp);
+  return int_power(static_cast<widen_t<T>>(bp.get_base()), power);
+}
+
+// A converter for the value member variable of magnitude (below).
+//
+// The input is the desired result, but in a (wider) intermediate type.  The point of this function
+// is to cast to the desired type, but avoid overflow in doing so.
+template<typename To, typename From>
+  requires std::is_arithmetic_v<To>
+    && std::is_arithmetic_v<From>
+    && (std::is_integral_v<To> == std::is_integral_v<From>)
+consteval To checked_static_cast(From x) {
+  // This function can only ever be called at compile time.  The purpose of these exceptions is to
+  // produce compiler errors, because we cannot `static_assert` on function arguments.
+  if constexpr (std::is_integral_v<To>) {
+    if (std::cmp_less(x, std::numeric_limits<To>::min()) ||
+        std::cmp_greater(x, std::numeric_limits<To>::max())) {
```

Comment:
```suggestion
    if (std::in_range<To>(x)) {
```

Reply:
Awesome---I didn't know about this!

Done (although it needed to be `!in_range`).